backtesting/backtesting.py

```diff
@@ -1420,9 +1420,12 @@ def _optimize_skopt() -> Union[pd.Series,
 
             # np.inf/np.nan breaks sklearn, np.finfo(float).max breaks skopt.plots.plot_objective
             INVALID = 1e300
+            skopt_pbar = _tqdm(dimensions, total=max_tries, desc="Skopt optimizations")
 
             @use_named_args(dimensions=dimensions)
             def objective_function(**params):
+                if hasattr(skopt_pbar, "update"):
```

Comment:
When does the progress bar not have '.update()`?
https://github.com/tqdm/tqdm/blob/bcce20f771a16cb8e4ac5cc5b2307374a2c0e535/tqdm/std.py#L1195-L1207

I don't think the [CI failures](https://github.com/kernc/backtesting.py/runs/2160506802) were related (:cold_sweat:). Prior commit was the better one. And we can't pass [`dimensions` as the iterable](https://github.com/tqdm/tqdm#documentation); it's `max_tries` that counts.

Reply:
It fails if `tqdm` is not installed and since it's not listed in the requirements or setup, it won't be installed by default: https://github.com/kernc/backtesting.py/blob/73e1534428fc2db9e681030862133b5067508520/backtesting/backtesting.py#L24-L29 

I can add it as a requirement and get rid of the import/attr checks. I assumed you're trying to keep requirements to a minimal.

Reply:
> I don't think the CI failures were related (cold_sweat). Prior commit was the better one. And we can't pass dimensions as the iterable; it's max_tries that counts.

Yup, `dimensions` was only used when tqdm isn't installed so that when the default method is executed, it has a value for `seq`. Otherwise it would error out.
```python
def _tqdm(seq, **_): 
         return seq 
```

By using `skopt_pbar.update(1)`, it no longer uses dimensions as the iterable. This way it accounted for both cases. I think it would be better if we explicitly add `tqdm` as a required package.

Reply:
Ah, right. I like being lightweight on dependencies. Even as tqdm is lightweight itself, I'd rather something else pulls it, and we just hook into using it.

The we should give our mock tqdm the missing method so as to rely on the uniform, tqdm interface. Simpler still, something like this might work:
```py
progress = iter(_tqdm(repeat(0), total=max_tries))

def objective(...):
    next(progress)
    ...
```